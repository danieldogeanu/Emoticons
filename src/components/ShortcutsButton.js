import React, { Component } from 'react';
import { animateCSS } from '../animate';
import Icon from './Icon';
import '../styles/components/ShortcutsButton.scss';

class ShortcutsButton extends Component {
	componentDidMount() {
		this.ShortcutsScreen = document.querySelector('.ShortcutsScreen');
		this.ShortcutsButton = document.querySelector('.ShortcutsButton');
	}

	handleClick() {
		animateCSS(this.ShortcutsScreen, 'fadeIn');
		this.ShortcutsScreen.classList.add('show');
		this.ShortcutsButton.classList.add('hide');
	}

	render() {
		const shortcutsButtonText = 'Keyboard Shortcuts';
		return (
			<button className="ShortcutsButton" 
				title={shortcutsButtonText}
				onClick={() => this.handleClick()}>
				<span className="show-for-screen-reader">
					{shortcutsButtonText}
				</span>
				<Icon name="keyboard" />
			</button>
		);
	}
}

export default ShortcutsButton;