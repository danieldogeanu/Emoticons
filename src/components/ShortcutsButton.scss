@import '../styles/includes';

.ShortcutsButton {
	display: block;
	position: fixed;
	left: rem(40);
	bottom: rem(30);
	width: auto;
	height: auto;
	border: 0 none;
	border-radius: rem(4);
	background-color: transparent;
	transition: color 0.3s ease, opacity 0.5s ease;
	color: color-scheme(scrollup, normal);
	cursor: pointer;
	&:hover { color: color-scheme(scrollup, hover); }
	&:focus, &:active { outline: none; }
	&.hide { display: none; }
	.icon.keyboard {
		width: rem(45);
		height: rem(45);
	}
}