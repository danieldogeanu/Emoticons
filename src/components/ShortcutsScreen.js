import React, { Component } from 'react';
import Icon from './Icon';
import './ShortcutsScreen.scss';

class ShortcutsClose extends Component {	
	handleClick() {
		const shortcutsScreen = document.querySelector('.ShortcutsScreen');
		const shortcutsButton = document.querySelector('.ShortcutsButton');
		shortcutsScreen.classList.remove('show');
		shortcutsButton.classList.remove('hide');
	}

	render() {
		const shortcutsCloseText = 'Close Shortcuts Screen';
		return (
			<button className="ShortcutsClose" 
				title={shortcutsCloseText}
				onClick={() => this.handleClick()}>
				<span className="show-for-screen-reader">
					{shortcutsCloseText}
				</span>
				<Icon name="close" />
			</button>
		);
	}
}

class ShortcutsScreen extends Component {
	render() {
		const shortcuts = [
			{ key: 's', description: 'Selects Search Input' },
			{ key: 'esc', description: 'Clears Search Input / Exits Menu' },
			{ key: 'k', description: 'Opens Keyboard Shortcuts Screen' },
			{ key: 't', description: 'Scrolls Back Up' },
		];
		return (
			<div className="ShortcutsScreen">
				<div className="shortcutsList">
					<span className="shortcutsTitle">Keyboard Shortcuts</span>
					<ul>
						{shortcuts.map(shortcut => {
							return <li key={shortcut.key}><kbd>{shortcut.key}</kbd> {shortcut.description}</li>;
						})}
					</ul>
					<ShortcutsClose />
				</div>
			</div>
		);
	}
}

export default ShortcutsScreen;