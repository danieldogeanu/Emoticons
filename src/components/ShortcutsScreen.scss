@import '../styles/includes';

// TODO: Add Animation for .ShortcutsScreen on Opening.

.ShortcutsScreen {
	display: none;
	position: fixed;
	top: 0; right: 0;
	bottom: 0; left: 0;
	padding: rem(30);
	@include linear-gradient(left, 
		color-scheme(shortcuts, gradient-start), rem(200), 
		color-scheme(shortcuts, gradient-end), 100%
	);
	@include breakpoint(min, rem(481)) { padding: rem(40); }
	&.show { display: block; }

	.shortcutsList {
		position: absolute;
		left: rem(30);
		bottom: rem(100);
		@include breakpoint(min, rem(481)) { left: rem(40); }
	}

	.shortcutsTitle {
		display: block;
		width: 100%;
		margin: 0 0 rem(20) 0;
		font-weight: weight(extra-bold);
		font-size: rem(14);
		text-transform: uppercase;
	}

	ul {
		display: block;
		list-style: none;
		margin: 0; padding: 0;
		li {
			margin: rem(12) 0;
			font-size: rem(16);
			&:hover kbd {
				border: rem(2) solid color-scheme(shortcuts, hover);
				color: color-scheme(shortcuts, hover);
			}
		}
	}

	kbd {
		display: inline-block;
		width: rem(45);
		margin: 0 rem(10) 0 0;
		padding: rem(6) rem(10);
		border-radius: rem(4);
		border: rem(2) solid color-scheme(shortcuts, normal);
		background-color: transparent;
		color: color-scheme(shortcuts, normal);
		font-family: font(open-sans);
		font-weight: weight(extra-bold);
		font-size: rem(12);
		line-height: 1;
		text-transform: uppercase;
		transition: all 0.3s ease;
		&:hover {
			border: rem(2) solid color-scheme(shortcuts, hover);
			color: color-scheme(shortcuts, hover);
		}
	}
}

.ShortcutsClose {
	display: flex;
	justify-content: flex-start;
	align-items: center;
	position: fixed;
	left: rem(30);
	bottom: rem(31);
	width: rem(45);
	height: rem(30);
	border: 0 none;
	border-radius: rem(4);
	background-color: transparent;
	transition: color 0.3s ease, opacity 0.5s ease;
	color: color-scheme(scrollup, normal);
	cursor: pointer;
	@include breakpoint(min, rem(481)) { left: rem(40); bottom: rem(41); }
	&:hover { color: color-scheme(scrollup, hover); }
	&:focus, &:active { outline: none; }
	.icon.close {
		width: rem(26);
		height: rem(26);
	}
}