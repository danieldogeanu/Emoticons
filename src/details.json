{
	"labels": [
		{
			"class": "face",
			"name": "Face"
		},
		{
			"class": "name",
			"name": "Emoticon Name"
		},
		{
			"class": "copy",
			"name": "Get"
		}
	],
	"author": {
		"name": "Daniel Dogeanu",
		"url": "https://danieldogeanu.com"
	},
	"social": [
		{
			"name": "Twitter",
			"url": "https://twitter.com/danieldogeanu"
		},
		{
			"name": "GitHub",
			"url": "https://github.com/danieldogeanu"
		},
		{
			"name": "LinkedIn",
			"url": "https://www.linkedin.com/in/danieldogeanu/"
		},
		{
			"name": "Behance",
			"url": "https://behance.net/danieldogeanu"
		}
	],
	"shortcuts": [
		{
			"key": "s",
			"description": "Selects Search Input"
		},
		{
			"key": "esc",
			"description": "Clears Search Input / Exits Menu"
		},
		{
			"key": "k",
			"description": "Opens Keyboard Shortcuts Screen"
		},
		{
			"key": "t",
			"description": "Scrolls Back Up"
		}
	]
}
