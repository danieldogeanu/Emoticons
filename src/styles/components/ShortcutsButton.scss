@import '../includes';

.ShortcutsButton {
	cursor: pointer;
	display: none;
	position: fixed;
	left: rem(30);
	bottom: rem(20);
	width: auto;
	height: auto;
	border: 0 none;
	border-radius: rem(4);
	background-color: transparent;
	color: color-scheme(scrollup, normal);
	transition: color 0.3s ease, opacity 0.5s ease;
	@include breakpoint(min, rem(481)) { left: rem(40); bottom: rem(30); }
	@include breakpoint(min, rem(961)) { display: block; }
	&:hover { color: color-scheme(scrollup, hover); }
	&:focus, &:active { outline: none; }
	&.hide { display: none; }
	.icon.keyboard {
		width: rem(45);
		height: rem(45);
	}
}

// TODO: Maybe include ShortcutsButton on a black bar at the bottom of the screen on mobile along with ScrollUp button.