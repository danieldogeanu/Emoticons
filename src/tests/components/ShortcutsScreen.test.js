import React from 'react';
import {render, fireEvent, cleanup, waitForDomChange} from '@testing-library/react';
import ShortcutsScreen from '../../components/ShortcutsScreen';
import ShortcutsButton from '../../components/ShortcutsButton';
import {shortcuts} from '../../details.json';

const screenTitle = 'Keyboard Shortcuts';
const openClasses = ['animated', 'fadeIn', 'show'];
const screenClasses = {
	container: 'shortcutsContainer',
	title: 'shortcutsTitle',
	list: 'shortcutsList',
	key: 'shortcutsKey',
};
const compNames = {
	button: 'ShortcutsButton',
	screen: 'ShortcutsScreen',
	close: 'ShortcutsClose',
	search: {
		bar: 'SearchBar',
		input: 'SearchInput',
	},
};

const compShell = (component) => (
	<div data-testid="test-parent">
		<form className={compNames.search.bar}
			data-testid={compNames.search.bar}>
			<input type="text" data-testid={compNames.search.input} />
		</form>
		<ShortcutsButton />
		{component}
	</div>
);

describe('ShortcutsScreen Component', () => {

	afterEach(cleanup);

	it('renders shortcuts screen properly', () => {
		const {getByTestId, getAllByTestId} = render(compShell(<ShortcutsScreen />));
		const renderedScreen = getByTestId(compNames.screen);
		const renderedContainer = getByTestId(screenClasses.container);
		const renderedTitle = getByTestId(screenClasses.title);
		const renderedList = getByTestId(screenClasses.list);
		const renderedKeys = getAllByTestId(screenClasses.key);

		expect(renderedScreen).toBeInTheDocument();
		expect(renderedScreen).toHaveClass(compNames.screen);
		expect(renderedScreen).toContainElement(renderedContainer);
		expect(renderedScreen).toContainElement(renderedTitle);
		expect(renderedScreen).toContainElement(renderedList);
		expect(renderedScreen).toContainElement(...renderedKeys);
		expect(renderedScreen).toMatchSnapshot();

		expect(renderedContainer).toHaveClass(screenClasses.container);
		expect(renderedTitle).toHaveClass(screenClasses.title);
		expect(renderedList).toHaveClass(screenClasses.list);

		expect(renderedTitle).toHaveTextContent(screenTitle);
		expect(renderedKeys.length).toBe(4);
	});

	it('renders all shortcut keys properly', () => {
		const {getAllByTestId} = render(compShell(<ShortcutsScreen />));
		const renderedKeys = getAllByTestId(screenClasses.key);

		renderedKeys.forEach((key, i) => {
			const kbd = key.querySelector('kbd');
			expect(kbd).toHaveTextContent(shortcuts[i].key);
			expect(key).toHaveTextContent(shortcuts[i].description);
			expect(key).toContainElement(kbd);
		});
	});

	it('opens on button click', () => {
		const {getByTestId, container} = render(compShell(<ShortcutsScreen />));
		const shortcutsScreen = getByTestId(compNames.screen);
		const shortcutsButton = container.querySelector(`.${compNames.button}`);

		fireEvent.click(shortcutsButton);

		expect(shortcutsScreen).toHaveClass(...openClasses);
	});

	it('closes on button click', async () => {
		const {getByTestId, container} = render(compShell(<ShortcutsScreen />));
		const shortcutsScreen = getByTestId(compNames.screen);
		const shortcutsButton = container.querySelector(`.${compNames.button}`);
		const shortcutsClose = container.querySelector(`.${compNames.close}`);

		fireEvent.click(shortcutsButton);
		fireEvent.click(shortcutsClose);

		await waitForDomChange();

		expect(shortcutsScreen).not.toHaveClass('show');
	});

	it('closes on \'esc\' key press', () => {
		const {getByTestId, container} = render(compShell(<ShortcutsScreen />));
		const shortcutsScreen = getByTestId(compNames.screen);
		const shortcutsButton = container.querySelector(`.${compNames.button}`);

		fireEvent.click(shortcutsButton);
		fireEvent.keyUp(shortcutsScreen, {
			key: 'Escape',
			code: 'Escape',
			keyCode: 27,
			charCode: 27
		});

		expect(shortcutsScreen).not.toHaveClass('show');
	});

	it('closes on \'k\' key press', () => {
		const {getByTestId, container} = render(compShell(<ShortcutsScreen />));
		const shortcutsScreen = getByTestId(compNames.screen);
		const shortcutsButton = container.querySelector(`.${compNames.button}`);

		fireEvent.click(shortcutsButton);
		fireEvent.keyUp(shortcutsScreen, {
			key: 'k',
			code: 'KeyK',
			keyCode: 75,
			charCode: 75
		});

		expect(shortcutsScreen).not.toHaveClass('show');
	});

});
