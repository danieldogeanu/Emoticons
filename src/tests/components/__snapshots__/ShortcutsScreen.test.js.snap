// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShortcutsScreen Component renders shortcuts screen properly 1`] = `
<div
  class="ShortcutsScreen"
  data-testid="ShortcutsScreen"
>
  <div
    class="shortcutsContainer"
    data-testid="shortcutsContainer"
  >
    <span
      class="shortcutsTitle"
      data-testid="shortcutsTitle"
    >
      Keyboard Shortcuts
    </span>
    <ul
      class="shortcutsList"
      data-testid="shortcutsList"
    >
      <li
        data-testid="shortcutsKey"
      >
        <kbd>
          s
        </kbd>
        Selects Search Input
      </li>
      <li
        data-testid="shortcutsKey"
      >
        <kbd>
          esc
        </kbd>
        Clears Search Input / Exits Menu
      </li>
      <li
        data-testid="shortcutsKey"
      >
        <kbd>
          k
        </kbd>
        Opens Keyboard Shortcuts Screen
      </li>
      <li
        data-testid="shortcutsKey"
      >
        <kbd>
          t
        </kbd>
        Scrolls Back Up
      </li>
    </ul>
    <button
      class="ShortcutsClose"
      title="Close Shortcuts Screen"
    >
      <span
        class="show-for-screen-reader"
        data-testid="ScreenReaderText"
      >
        Close Shortcuts Screen
      </span>
      <svg
        class="icon close"
      >
        <use
          xlink:href="#close"
        />
      </svg>
    </button>
  </div>
</div>
`;
